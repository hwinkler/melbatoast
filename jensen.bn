A 2
0.4 0.6


B|A 2
0.3 0.7 0.8 0.2


C|A 2
0.7 0.3 0.4 0.6


D|B 2
0.5 0.5 0.1 0.9

E|D,C 2
0.9 0.1 0.999 0.001
0.999 0.001 0.999 0.001

