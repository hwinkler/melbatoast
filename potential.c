#include <stdbool.h>
#include <stddef.h>
#include "potential.h"

bool _isParent(Potential* child, Potential* test){
  Potential* parent = NULL;
  for (int i = 0; 
       parent != NULL && i < MAX_PARENTS;
       i++){
    parent = child->parents[i];
    if (parent == test){
      return true;
    }
  }
  return false;
}

int sample (Potential* potential, State* states[],int numStates) {
  for (int i=0; i< numStates; i++){
    Potential* statePotential =  states[i]->potential;
    if (_isParent (potential, statePotential)){
      //statePotential is one of our parents
      
    } else if (_isParent (statePotential, potential)) {
      //statePotential is one of our children
      
        
    }
  }
  return -1;
}
