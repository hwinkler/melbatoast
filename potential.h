#pragma once

#define MAX_PARENTS 5
#define MAX_CHILDREN 5

typedef struct Potential {
  int   numStates;
  float *conditionals;
  int numConditionals;   // this is here mainly for debugging
  int numParents;
  int numChildren;
  struct Potential* parents[MAX_PARENTS];
  struct Potential* children[MAX_CHILDREN];
  int indexInChild [MAX_CHILDREN];
  int dimensions [MAX_PARENTS + 1];

  int state;  /// TODO remove. This is the only mutable field.
  int isFrozen;

} Potential;

