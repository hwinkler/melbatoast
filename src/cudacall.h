#pragma once

#define CUDA_CALL(x) do { if((x) != cudaSuccess) { \
      printf("Error %d at %s:%d;  %s\n",x, __FILE__,__LINE__,   cudaGetErrorString(x)); \
      return EXIT_FAILURE;}} while(0)
