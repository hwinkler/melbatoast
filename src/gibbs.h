#pragma once

__global__
void initPotential(Potential*p, 
                    int numStates,
                    float* conditionals,
                    Potential** parents,
                   int numParents);

__global__
void gibbs ( const Potential* __restrict__ potentials, int numPotentials, const int *__restrict__  initialStates,
            int countsBase[], int numCounts, int numIterations);
