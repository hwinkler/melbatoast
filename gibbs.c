#include <stdbool.h>
#include <stddef.h>
#include <string.h>
#include <assert.h>
#include <stdio.h>

#include "potential.h"
#include "projection.h"

#ifndef DEBUG
#define DEBUG 0
#endif

#define DPRINT(...)                                             \
  do { if (DEBUG) fprintf(stderr, __VA_ARGS__); } while (0)


// TODO: including mtwist.h caused duplicate defs in the linker
extern double mt_drand();
extern void mt_seed();

void _initDistribution (float* distribution, int n){
  for (int i=0; i< n; i++){
    distribution[i] = 1.0f;
  }
}
void _normalizeDistribution (float* distribution, int n){
  float sum = 0.f;
  for (int i=0; i< n; i++){
    assert(distribution[i] > 0.f);
    sum += distribution[i] ;
  }
  float scale = 1.f / sum;
  for (int i=0; i< n; i++){
    distribution[i] *= scale ;
  }
}

void _cumulativeDistribution (float* distribution, float * cumulative, int n){
  cumulative[0] = distribution[0];
  for (int i=1; i< n; i++){
    cumulative[i] = cumulative[i-1] + distribution[i] ;
  }
}

int _drawFromCumulative (float* cumulative, int n){
  double r =  mt_drand();
  
  for (int i=0; i<n; i++){
    
    if ( r < cumulative[i]){
      return i;
    }
  }
  
  assert (false);
  return 0;
}

void _conditionalGiven(Potential *potential ,int indexUnfixed, float* distribution){

  int numDimensions = potential->numParents + 1;
  int indices[numDimensions];

  indices[0] = potential->state;
  for (int iParent=0; iParent < potential->numParents; iParent++){
    assert (iParent+1 < numDimensions);
    indices[iParent+1] = potential->parents[iParent]->state;
  }

  assert (indexUnfixed < numDimensions);
  indices[indexUnfixed] = -1;

  int offsetOut = 0, lengthOut =0, strideOut =0;
  projection ( potential->dimensions, indices, numDimensions,
               &offsetOut, &lengthOut, &strideOut);
  assert (lengthOut == potential->numStates);
  
  for (int iState =0; iState < potential->numStates; iState++){
    assert (offsetOut + iState * strideOut < potential->numConditionals);
    distribution[iState] *= potential->conditionals[offsetOut + iState * strideOut];

    
  }
  
}

void _printArray(float * f, int n){
  for (int i=0; i<n; i++){
    
  }
  
}


void gibbs (Potential** potentials, int numPotentials, 
            int counts[], int numCounts, int numIterations) {
  
  mt_seed();

  memset (counts, 0, numCounts* sizeof(int));
  
  for (int i=0; i<numIterations; i++){
    
    

    for (int j=0; j < numPotentials; j++){
      Potential *p = potentials[j];
      if (p->isFrozen){
        continue;
      }
      float distribution [p->numStates];
      _initDistribution(distribution, p->numStates);

      
     
      // Obtain the conditional distribution for the current potential

      Potential* potential = p;

      _conditionalGiven (p, 0, distribution);

      
      
      // Multiply in the distribution for this variable in each child potential
      for (int iChild =0; iChild < p->numChildren; iChild++){
        Potential * child = p->children[iChild];
     
        // add one to indexInChild, since the indexInChild refers to zero based
        // among the parents -- but zer oindex is reserved to the potential's variable
        _conditionalGiven (child, p->indexInChild[iChild] + 1, distribution);
        
        _printArray(distribution, p->numStates);
      }
      
      _normalizeDistribution (distribution, p->numStates);
      
      _printArray(distribution, p->numStates);

      float cumulative [p->numStates];
      _cumulativeDistribution (distribution, cumulative, p->numStates);

      int newState = _drawFromCumulative(cumulative, p->numStates);

      p->state = newState;
    }

    // which configuration is this?

    int config = 0;
    for (int j=0; j < numPotentials; j++){
      config *= potentials[j]->numStates;
      config += potentials[j]->state;
    }
    
    counts[config] ++;
  }

}
