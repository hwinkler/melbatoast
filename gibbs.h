#pragma once
#include "potential.h"
extern void gibbs (const Potential* const potentials, int numPotentials, const int * const initialStates,
                   int counts[], int numCount, int numIterations);

