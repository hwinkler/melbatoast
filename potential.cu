#include <string.h>
#include <assert.h>
#include <stdbool.h>
#include <stdio.h>
#include "potential.h"



__global__ void add(int *a, int *b, int *c) { 
  *c = *a + *b;
}

