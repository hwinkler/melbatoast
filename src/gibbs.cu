#include <curand_kernel.h>
#include <assert.h>
#include "constants.h"
#include "potential.h"

#define DPRINT(...)

__device__ double rnd(curandState* state) {
  return curand_uniform( state );
}

__device__ void rndSeed ( curandState * state){
  unsigned int seed = (unsigned int) clock();
  int id =  threadIdx.x + blockIdx.x * blockDim.x;
  curand_init ( seed, id, 0, state );
} 


__device__
void projection (
                 const int *const dimensions,
                 const int *const indices,
                 int numDimensions,
                 int *offsetOut,
                 int *lengthOut,
                 int *strideOut) {

  // the first dimension varies fastest
  // index -1 indicates the dimension we are extracting

  int hypersheetSize = 1;
  int stride = -1;
  int index0 = 0;
  int length = -1;
  
  for (int i =0; i < numDimensions ; i++){

    int index;
    if (indices[i] < 0){
      index = 0;
      stride = hypersheetSize;
      length = dimensions[i];
    } else {
      index = indices[i];
    }
    index0 += hypersheetSize * index;
    hypersheetSize *= dimensions[i];
  }

  *offsetOut = index0;
  *lengthOut = length;
  *strideOut = stride;

}


__device__
void _initDistribution (float* distribution, int n){
  for (int i=0; i< n; i++){
    distribution[i] = 1.0f;
  }
}

__device__
void _normalizeDistribution (float* distribution, int n){
  float sum = 0.f;
  for (int i=0; i< n; i++){
    assert(distribution[i] > 0.f);
    sum += distribution[i] ;
  }
  float scale = 1.f / sum;
  for (int i=0; i< n; i++){
    distribution[i] *= scale ;
  }
}

__device__
void _cumulativeDistribution (float* distribution, float * cumulative, int n){
  cumulative[0] = distribution[0];
  for (int i=1; i< n; i++){
    cumulative[i] = cumulative[i-1] + distribution[i] ;
  }
}

__device__
int _drawFromCumulative (float* cumulative, int n, curandState* rndState){
  double r =  rnd(rndState);
  
  for (int i=0; i<n; i++){
   
    if ( r <= cumulative[i]){
      return i;
    }
  }
  //DPRINT("drawFromCumulative %d %f %f\n", n, r, cumulative[n-1]);
  //assert (false);
  return n-1;
}

__device__
void _conditionalGiven(const Potential *const potentials , int offset, const int * const states, int indexUnfixed, float* distribution){

  const Potential *const potential = potentials + offset;
  int numDimensions = potential->numParents + 1;
  int indices[MAX_DIMENSIONS];

  indices[0] = states[offset];
  for (int iParent=0; iParent < potential->numParents; iParent++){
    assert (iParent+1 < numDimensions);
    Potential *parent = potential->parents[iParent];
    int parentOffset = parent - potentials;
    indices[iParent+1] = states[parentOffset];
  }

  assert (indexUnfixed < numDimensions);
  indices[indexUnfixed] = -1;

  int offsetOut = 0, lengthOut =0, strideOut =0;
  projection ( potential->dimensions, indices, numDimensions,
               &offsetOut, &lengthOut, &strideOut);
  assert (lengthOut == potential->numStates);
  
  for (int iState =0; iState < potential->numStates; iState++){
    assert (offsetOut + iState * strideOut < potential->numConditionals);
    distribution[iState] *= potential->conditionals[offsetOut + iState * strideOut]; 
  }
}

__global__
void gibbs (const Potential* const potentials, int numPotentials, const int *const initialStates,
            int countsBase[], int numCounts, int numIterations) {
  curandState rndState;
  rndSeed(&rndState);

  if (numPotentials > MAX_POTENTIALS){
    return;
  }
  int states[MAX_POTENTIALS]; 
  memcpy (states, initialStates, numPotentials*sizeof(int));
  
  // int * counts = countsBase + blockIdx.x * numCounts;
  int counts[MAX_CONFIGURATIONS];
  memset ( counts, 0, numCounts* sizeof(int));
    
  for (int i=0; i<numIterations; i++){
    
    for (int j=0; j < numPotentials; j++){
      const Potential *const p = potentials + j;
      if (p->isFrozen){
        continue;
      }
      
      float distribution [MAX_STATES];
      _initDistribution(distribution, p->numStates);

           
      // Obtain the conditional distribution for the current potential

      _conditionalGiven (potentials, j, states, 0, distribution);     
      
      // Multiply in the distribution for this variable in each child potential
      for (int iChild =0; iChild < p->numChildren; iChild++){
        Potential * child = p->children[iChild];
     
        // add one to indexInChild, since the indexInChild refers to zero based
        // among the parents -- but zer oindex is reserved to the potential's variable
        _conditionalGiven (potentials, child - potentials, states, p->indexInChild[iChild] + 1, distribution);
        
      }
      
      _normalizeDistribution (distribution, p->numStates);
      
      float cumulative [MAX_STATES];
      _cumulativeDistribution (distribution, cumulative, p->numStates);

      int newState = _drawFromCumulative(cumulative, p->numStates, &rndState);

      states[j] = newState;
    }

    // which configuration is this?

    int config = 0;
    for (int j=0; j < numPotentials; j++){
      config *= potentials[j].numStates;
      config += states[j];
    }
    
    counts[config] ++;
  }

  int * gCounts = countsBase ;//+ blockIdx.x * numCounts;
  for (int i=0; i<numCounts; i++){
    atomicAdd(gCounts+i, counts[i]);
  }

}

__global__
void initPotential(Potential*p, 
                    int numStates,
                    float* conditionals,
                    Potential** parents,
                    int numParents) {
  p->numStates = numStates;
  int numConditionals = numStates;

  for (int iParent=0; iParent < numParents; iParent++){
    Potential * parent = parents[iParent];

    numConditionals *= parent->numStates;
    int indexOfChildInParent = parent->numChildren;
    parent->children[indexOfChildInParent] = p;
    parent->indexInChild[indexOfChildInParent] = iParent;
    parent->numChildren++;
  }


  p->numConditionals = numConditionals;
  p->conditionals = conditionals;

  memset (p->parents, 0, sizeof(Potential*) * MAX_PARENTS);
  memset (p->children, 0, sizeof(Potential*) * MAX_CHILDREN);
  memset (p->indexInChild, 0, sizeof(int) * MAX_CHILDREN);

  p->numParents = numParents;
  p->numChildren = 0;
  if (numParents > 0){ //suspicious that we need this or else mem err
    memcpy (p->parents,  parents, numParents * sizeof(Potential*));
  } 
  
  int numDimensions = 1 + numParents;
  p->dimensions[0] = numStates;
  for (int iDim = 1; iDim < numDimensions; iDim++){
    Potential *parent = p->parents[iDim-1];
    
    p->dimensions[iDim] = parent->numStates;
  }
  p->isFrozen = false;
}
