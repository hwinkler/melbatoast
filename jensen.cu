#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <cuda.h>
#include "potential.h"
#include "gibbs.h"
#include "cudacall.h"

int printDevicePotential (Potential*pd) {
  Potential p;  
  CUDA_CALL(cudaMemcpy ( &p,  pd, sizeof(p), cudaMemcpyDeviceToHost));

  printf("%17s %6d\n",  "numStates", p.numStates);

  if (p.numConditionals >=0 && p.numConditionals < 1000){
    float conditionals[p.numConditionals];
    CUDA_CALL(cudaMemcpy ( conditionals,  p.conditionals, p.numConditionals, cudaMemcpyDeviceToHost));
    for (int i=0; i< p.numConditionals; i++){
      printf("%11s[%3d] %6.3f\n",  "conditionals", i,  p.conditionals[i]);
    }
  }
  printf("%17s %6d\n",  "numConditionals", p.numConditionals);
  printf("%17s %6d\n",  "numParents", p.numParents);
  printf("%17s %6d\n",  "numChildren", p.numChildren);
  

  if (p.numParents >=0 && p.numParents <= MAX_PARENTS){
    for (int i=0; i<p.numParents; i++){
      int offset = p.parents[i] - pd;
      printf("%11s[%3d] %6d\n",  "parent", i, offset);
    }
  }
  if (p.numChildren >=0 && p.numChildren <= MAX_CHILDREN){
    for (int i=0; i<p.numChildren; i++){
      int offset = p.children[i] - pd;
      printf("%11s[%3d] %6d\n",  "child", i, offset);
    }
  }
  if (p.numChildren >=0 && p.numChildren <= MAX_CHILDREN){
    for (int i=0; i<p.numChildren; i++){
      printf("%11s[%3d] %6d\n",  "indexInChild", i, p.indexInChild[i]);
    }
  }
  if (p.numParents >=0 && p.numParents <= MAX_PARENTS){
    for (int i=0; i<=p.numParents; i++){
      printf("%11s[%3d] %6d\n",  "dimensions", i, p.dimensions[i]);
    }
  }
  printf("%17s %6d\n",  "isFrozen", p.isFrozen);
  return 0;
}


  __global__ void add(int *a, int *b, int *c);
int main(void) {

  int a, b, c; // host copies of a, b, c 
  int *d_a, *d_b, *d_c; // device copies of a, b, c 
  int size = sizeof(int);
  // Allocate space for device copies of a, b, c
  cudaMalloc((void **)&d_a, size); 
  cudaMalloc((void **)&d_b, size); 
  cudaMalloc((void **)&d_c, size);
  // Setup input values
  a = 2; b = 7;
  cudaMemcpy(d_a, &a, size, cudaMemcpyHostToDevice);
  cudaMemcpy(d_b, &b, size, cudaMemcpyHostToDevice);
  // Launch add() kernel on GPU
  add<<<1,1>>>(d_a, d_b, d_c);
  // Copy result back to host
  cudaMemcpy(&c, d_c, size, cudaMemcpyDeviceToHost);
  // Cleanup
  cudaFree(d_a); cudaFree(d_b); cudaFree(d_c); 


  printf("%d + %d = %d\n", a,b,c);
  return 0;
}

int mainXXX (int argc, char ** argv){

  int N = 1;

  const int numPotentials = 5;
  Potential* devPotentials;
  CUDA_CALL(cudaMalloc ( (void**) &devPotentials, numPotentials *sizeof( Potential ) ));
  Potential * da = devPotentials+0, *db = devPotentials+1, *dc=devPotentials+2, *dd=devPotentials+3, *de = devPotentials+4;


  const int numConditionals = 2+4+4+4+8 ;
  float conditionals[numConditionals] = {
    // a
    0.4f, 0.6f,
    //b
    0.3f, 0.7f, 0.8f, 0.2f,
    //c
    0.7f, 0.3f, 0.4f, 0.6f,
    //d
    0.5f, 0.5f, 0.1f, 0.9,
    // e
    0.9f, 0.1f, 0.999f,0.001f,
    0.999f, 0.001f,  0.999f,0.001f};

  float *devConditionals;
  CUDA_CALL(cudaMalloc ( (void**) &devConditionals, numConditionals * sizeof( float ) ));
  CUDA_CALL(cudaMemcpy (devConditionals, conditionals, numConditionals* sizeof(float), cudaMemcpyHostToDevice));
  float *dca = devConditionals+0, *dcb = dca + 2, *dcc = dcb +4, *dcd = dcc + 4, *dce = dcd + 4;

  const int numStates[numPotentials] = {2,2,2,2,2};
  // P(A)
  initPotential<<<1, 1>>> (da, numStates[0], dca, 
                           (Potential *[]) {NULL}, 0 );


  // P(B|A)
  initPotential<<<1, 1>>> (db, numStates[1], dcb, 
                           (Potential *[]) {da}, 1 );
  // P(C|A)
  initPotential<<<1, 1>>> (dc, numStates[2], dcc, 
                           (Potential *[]) {da}, 1 );
  // P(D|B)
  initPotential<<<1, 1>>> (dd, numStates[3], dcd, 
                           (Potential *[]) {db}, 1 );
  // P(E|D,C)
  initPotential<<<1, 1>>> (de, numStates[4], dce,
                           (Potential *[]) {dd, dc}, 2 );
  
  
  for (int i=0; i< numPotentials; i++){
    Potential* p = da + i;
    printf ("Potential %c %p:\n", 'A' + i, p);
    printDevicePotential(p);
  }

  //data: B=n, E=n
 
 
  // b->isFrozen = e->isFrozen = true;
  
  int numConfigurations = 1;
  for (int i=0; i< numPotentials; i++){
    numConfigurations *= numStates[i];
  }
  int counts[numConfigurations];
  memset (counts, 0, numConfigurations * sizeof(int));

  // initial config: ynyyn  (we use y=0, n=1)
  int  states [numPotentials] = {0,1,0,0,1};
  int * devStates;
  CUDA_CALL(cudaMalloc((void**)&devStates,  numPotentials* sizeof(int)));
  CUDA_CALL(cudaMemcpy (devStates, states, numPotentials* sizeof(int), cudaMemcpyHostToDevice));

  int * devCounts ;
  CUDA_CALL(cudaMalloc( (void**) &devCounts, numConfigurations* sizeof(int)));
  CUDA_CALL(cudaMemcpy (devCounts, counts, numConfigurations* sizeof(int), cudaMemcpyHostToDevice));

  gibbs<<<1,N>>>(devPotentials, numPotentials, devStates, devCounts, numConfigurations, 100);

  CUDA_CALL(cudaMemcpy ( counts,  devCounts, numConfigurations* sizeof(int), cudaMemcpyDeviceToHost));

  for (int j=0; j < numConfigurations; j++){
    printf("%4d: %4d\n", j, counts[j]);
  }


  return 0;
}
