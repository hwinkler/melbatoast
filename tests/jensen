#!/bin/bash

DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
$DIR/../gibbs -f "$DIR/jensen.bn" -s "$DIR/jensen.state" > "$DIR/jensen.out"

