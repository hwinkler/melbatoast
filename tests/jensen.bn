A 
't 'f
0.4 0.6


B|A 
't 'f
0.3 0.7 0.8 0.2


C|A
't 'f
0.7 0.3 0.4 0.6


D|B
't 'f
0.5 0.5 0.1 0.9

E|D,C 
't 'f
0.9 0.1 0.999 0.001
0.999 0.001 0.999 0.001

