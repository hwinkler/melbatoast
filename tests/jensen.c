#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include "../potential.h"
#include "../gibbs.h"

int main (int argc, char ** argv){

  const int numConditionals = 2+4+4+4+8 ;
  float conditionals[numConditionals];
  Potential potentials [5]; //causal order
  Potential * a = potentials+0, *b = potentials+1, *c=potentials+2, *d=potentials+3, *e = potentials+4;


  float *ca = conditionals+ 0;
  memcpy (ca,( (float []){0.4f, 0.6f}), 2 * sizeof(float));
  float *cb = ca + 2;
  memcpy (cb,( (float []){0.3f, 0.7f, 0.8f, 0.2f}), 4 * sizeof(float));
  float *cc = cb + 4;
  memcpy (cc,( (float []){0.7f, 0.3f, 0.4f, 0.6f}), 4 * sizeof(float));
  float *cd = cc + 4;
  memcpy (cd,( (float []){0.5f, 0.5f, 0.1f, 0.9}), 4 * sizeof(float));
  float *ce = cd + 4;
  memcpy (ce,( (float []) {
        0.9f, 0.1f, 0.999f,0.001f,
          0.999f, 0.001f,  0.999f,0.001f}), 8 * sizeof(float));

  // P(A)

  initPotential (a, 2, ca, 
                  (Potential *[]) {NULL}, 0 );

  
 
  // P(B|A)
  initPotential (b, 2, cb, 
                  (Potential *[]) {a}, 1 );
  

  // P(C|A)
  initPotential (c, 2, cc, 
                  (Potential *[]) {a}, 1 );
   

  // P(D|B)
  initPotential (d, 2, cd, 
                  (Potential *[]) {b}, 1 );
  

  // P(E|D,C)
  initPotential (e, 2, ce,
                  (Potential *[]) {d, c}, 2 );
  

  //data: B=n, E=n
  // initial config: ynyyn  (we use y=0, n=1)

  int  states [5] = {0,1,0,0,1};

  // b->isFrozen = e->isFrozen = true;

  const int numPotentials = 5;


  int numPossibleConfigurations = 1;
  for (int i=0; i< numPotentials; i++){
    numPossibleConfigurations *= potentials[i].numStates;
  }
  int counts[numPossibleConfigurations];

  gibbs(potentials, numPotentials, states, counts, numPossibleConfigurations, 1000000);


  for (int j=0; j < numPossibleConfigurations; j++){
    printf("%4d: %4d\n", j, counts[j]);
  }

}
