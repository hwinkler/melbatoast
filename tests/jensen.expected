   0:    0
   1:    0
   2:    0
   3:    0
   4:    0
   5:    0
   6:    0
   7:    0
   8:    0
   9: 695601
  10:    0
  11: 62748
  12:    0
  13: 2915
  14:    0
  15: 26926
  16:    0
  17:    0
  18:    0
  19:    0
  20:    0
  21:    0
  22:    0
  23:    0
  24:    0
  25: 170690
  26:    0
  27: 15495
  28:    0
  29: 2502
  30:    0
  31: 23123
